dynamiqs/smesolve/smesolve.py

```diff
@@ -1,23 +1,105 @@
-        'back soon!'
+    *,
+    t_meas: TensorLike | None = None,
+    seed: int | None = None,
+    exp_ops: OperatorLike | list[OperatorLike] | None = None,
+    options: Options | None = None,
+) -> tuple[Tensor, Tensor | None, Tensor | None]:
+    # H: (b_H?, n, n), rho0: (b_rho0?, n, n) -> (y_save, exp_save, meas_save) with
+    #    - y_save: (b_H?, b_rho0?, ntrajs, len(t_save), n, n)
+    #    - exp_save: (b_H?, b_rho0?, ntrajs, len(exp_ops), len(t_save))
+    #    - meas_save: (b_H?, b_rho0?, ntrajs, len(meas_ops), len(t_meas) - 1)
+
+    # check jump_ops
+    if isinstance(jump_ops, list) and len(jump_ops) == 0:
+        raise ValueError(
+            'Argument `jump_ops` must be a non-empty list of tensors. Otherwise,'
+            ' consider using `ssesolve`.'
+        )
+
+    # format and batch all tensors
+    formatter = TensorFormatter(options.dtype, options.device)
+    H_batched, rho0_batched = formatter.format_H_and_state(H, rho0, state_to_dm=True)
+    H_batched = H_batched.unsqueeze(2)
+    rho0_batched = rho0_batched.unsqueeze(2).repeat(1, 1, ntrajs, 1, 1)
+    # H_batched: (b_H, 1, 1, n, n)
+    # rho0_batched: (b_H, b_rho0, ntrajs, n, n)
+    exp_ops = formatter.format(exp_ops)  # (len(exp_ops), n, n)
+    jump_ops = formatter.format(jump_ops)  # (len(jump_ops), n, n)
+
+    # convert t_save to a tensor
+    t_save = torch.as_tensor(t_save, dtype=options.rdtype, device=options.device)
+
+    # convert etas to a tensor and check
+    etas = torch.as_tensor(etas, dtype=options.rdtype, device=options.device)
+    if len(etas) != len(jump_ops):
+        raise ValueError('Argument `etas` must have the same length as `jump_ops`.')
+    if torch.all(etas == 0.0):
+        raise ValueError(
+            'Argument `etas` must contain at least one non-zero value. Otherwise, '
+            'consider using `mesolve`.'
+        )
+    if torch.any(etas < 0.0) or torch.any(etas > 1.0):
+        raise ValueError('Argument `etas` must contain values between 0 and 1.')
+
+    # split jump operators between purely dissipative (eta = 0) and monitored (eta != 0)
+    mask = etas == 0.0
+    meas_ops, etas = jump_ops[~mask], etas[~mask]
+
+    # convert t_meas to a tensor
+    t_meas = torch.as_tensor(
+        [] if t_meas is None else t_meas, dtype=options.rdtype, device=options.device
     )
+
+    # define random number generator from seed
+    generator = torch.Generator(device=options.device)
+    generator.seed() if seed is None else generator.manual_seed(seed)
+
+    # default options
+    if options is None:
+        raise ValueError(
+            'No default solver yet, please specify one using the options argument.'
+        )
+
+    # define the solver
+    args = (H_batched, rho0_batched, exp_ops, options)
+    kwargs = dict(
+        jump_ops=jump_ops,
+        meas_ops=meas_ops,
+        etas=etas,
+        generator=generator,
+        t_save_y=t_save,
```

Comment:
A general comment regarding the PR (I'm putting it here but it's PR-wide). 

I don't quite understand why you introduced `t_save_y` on top of the `t_save` we had before. It seems you initialize it to `t_save` everywhere anyway (in `smesolve`, in `mesolve` and in `sesolve`). Plus it complexifies the way you handle saving, and not only for `smesolve` but for all Solver child classes.

If you need a solution specificaly for `smesolve`, I think we should implement a solution that does not alter the other Solvers (unless we'll need it as well at some point, but it doesn't look like it for now).

Reply:
So in short, if there's a way to simplify this whole saving workflow, I think it'd be nice and might avoid us problems later on.

Reply:
`t_save` defined by the `Solver` base class is "the times when smthg has to be saved", and then we have `t_save_y` to save the state and exp_ops, and `t_save_meas` to save the measurement record (for example for the SME, `t_save = torch.cat(t_save_y, t_save_meas)`). This is necessary to
- specify correct tensor dimensions when we initialize y_save and exp_save
- differentiate when we save the state vs when we save the measurement record

Ideally, we would have defined a base solver class with a `t_save` that doesn't know what to save, just when something has to be saved. Then we would inherit to define the logic to save states `t_save_y`, and inherit again to define the logic to save measurement record `t_save_meas`. The confusion comes because we directly implement the saving of the state and exp_ops in the base solver class. If you prefer I can write this base class that just knows about `t_save` and knows nothing about the problem, but I thought it's a bit overkill.

Reply:
If you have a better idea I'd be happy to try! But I think the best solution is to make the `Solver` base class a bit more generic concerning saving.

Reply:
Ok thanks for the clarification it's clearer ! I didn't understand that you meant `t_save` for any save time, and I thought `t_save_y` was just the times at which we save the state (and not the exp_ops). 

Maybe what I would propose is the following:

- keep the overall implementation you propose
- rename `t_save_y` to `t_save` as we had before. Currently, `t_save` is a different object when passed by users and within the code. Also, the `_y` was misleading IMO.
- rename `t_save` to `t_allsave` (or similar) to account for all save times (i.e. the concatenation of `t_save` and `t_meas`).
- in the arguments of the `Solver` base class, pass only `t_save`, and add a method `init_t_allsave` (or similar) that computes the `t_allsave` object from `t_save` and `t_meas`. There will be one such method in the base class, and one in the `SMESolver` class.
- in the definition of `SMESolver._save`, avoid if possible the `if t in self.t_save_y` lines by using a counter and some pre-processing. It's nice not to search through the object, and we should know in advance when to save y and when to save the measurements. It's a bit faster and a bit more robust.

What do you think?

Reply:
I tried something, it's quite complicated but I don't know how we could simplify it 😅 Any suggestions welcome!